src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/SourceCodeUpdater.cs

```diff
@@ -0,0 +1,285 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater
+{
+    public class SourceCodeUpdater
+    {
+        private readonly SyntaxTree _template;
+        private readonly SyntaxTree _programFile;
+        private readonly ILogger<SourceCodeUpdater> _logger;
+
+        public SourceCodeUpdater(SyntaxTree sourceCode, string template, ILogger<SourceCodeUpdater> logger)
+        {
+            _programFile = sourceCode;
+            _logger = logger;
+            template = ReplaceNames(sourceCode.GetRoot(), template);
+            _template = CSharpSyntaxTree.ParseText(template);
+        }
+
+        // Simple constructor for source code files which only needs to update the using directives
+        public SourceCodeUpdater(SyntaxTree sourceCode, ILogger<SourceCodeUpdater> logger)
+        {
+            _programFile = sourceCode;
+            _template = CSharpSyntaxTree.ParseText(string.Empty);
+            _logger = logger;
+        }
+
+        // Wraps different steps of source code update in one method
+        public SyntaxNode SourceCodeUpdate()
+        {
+            var root = UpdateDirectives();
+            root = AddTemplateCode(root);
+            return RemoveOldCode(root);
+        }
+
+        // Updates the using directives by deleting ServiceModel directives and adding CoreWCF directives
+        public SyntaxNode UpdateDirectives()
+        {
+            if (_programFile is null)
+            {
+                throw new ArgumentNullException(nameof(_programFile));
+            }
+
+            var root = _programFile.GetRoot();
+
+            // removes old directives
+            var oldDirectives = from r in root.DescendantNodes().OfType<UsingDirectiveSyntax>()
+                                where ContainsIdentifier("ServiceModel", r)
+                                select r;
+            root = root.RemoveNodes(oldDirectives, 0)!;
+
+            // adds new directives and avoid duplicate using directives
+            var template = Constants.TemplateUsing;
+            if (string.IsNullOrEmpty(_template.GetText().ToString()))
+            {
+                template = Constants.TemplateUsingShort;
+            }
+
+            var currDirectives = from d in root.DescendantNodes().OfType<UsingDirectiveSyntax>()
+                                 where template.Contains(d.ToFullString(), StringComparison.Ordinal)
+                                 select d.ToFullString().Replace(System.Environment.NewLine, string.Empty);
+            var result = string.Empty;
```

Comment:
I think d.Trim() might be better. There are a couple of potential issues with this. The NewLine character for the currently running OS might be different than what's in the file. Linux and Windows have a different NewLine sequence. You could have a combination which leaves a character behind which would mean it wouldn't match the line on the Contains call a few lines below. Also some people have extra blank lines at the end of the line before the new line and the current code would break on that. I belied string.Trim() will cover all those cases.